lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -16,18 +16,311 @@
+  /// debugging purposes.
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &os, MutableNLA &x) {
+    auto writePathSegment = [&](StringAttr mod, StringAttr sym = {}) {
+      if (sym)
+        os << "#hw.innerNameRef<";
+      os << "@" << mod.getValue();
+      if (sym)
+        os << "::@" << sym.getValue() << ">";
+    };
+
+    auto writeOne = [&](StringAttr root, StringAttr sym) {
+      os << "firrtl.nla @" << sym.getValue() << " [";
+
+      StringAttr lastMod;
+      // Split out the first iteration to simplify loop logic.
+      if (!x.inlinedSymbols.test(1))
+        lastMod = root;
+      else
+        writePathSegment(root, x.nla.refPart(0));
+      // Rest of the loop.
+      bool needsComma = false;
+      for (signed i = 1, e = x.inlinedSymbols.size() - 1; i != e; ++i) {
+        if (i == x.flattenPoint) {
+          lastMod = x.nla.modPart(i);
+          break;
+        }
+
+        if (!x.inlinedSymbols.test(i + 1)) {
+          if (!lastMod)
+            lastMod = x.nla.modPart(i);
+          continue;
+        }
+
+        if (needsComma)
+          os << ", ";
+        writePathSegment(lastMod ? lastMod : x.nla.modPart(i),
+                         x.nla.refPart(i));
+        needsComma = true;
+        lastMod = {};
+      }
+
+      os << ", ";
+      writePathSegment(lastMod ? lastMod : x.nla.modPart(x.size - 1),
+                       x.nla.refPart(x.size - 1));
+      os << "]";
+    };
+
+    SmallVector<InnerRefAttr> tops;
+    if (!x.dead)
+      tops.push_back(InnerRefAttr::get(x.nla.root(), x.nla.getNameAttr()));
+    tops.append(x.newTops.begin(), x.newTops.end());
+
+    bool multiary = !x.newTops.empty();
+    if (multiary)
+      os << "[";
+    llvm::interleaveComma(tops, os, [&](InnerRefAttr a) {
+      writeOne(a.getModule(), a.getName());
+    });
+    if (multiary)
+      os << "]";
+
+    return os;
+  }
+
+  /// Returns true if this NLA is dead.  There are several reasons why this
+  /// could be dead:
+  ///   1. This NLA has no uses and was not re-topped.
+  ///   2. This NLA was flattened and its leaf reference is a Module.
+  bool isDead() { return dead && newTops.empty(); }
+
+  /// Returns true if this NLA is local.  For this to be local, every module
+  /// after the root (up to the flatten point or the end) must be inlined.  The
+  /// root is never truly inlined as inlined as inlining the root just sets a
+  /// new root.
+  bool isLocal() {
+    unsigned end = flattenPoint > -1 ? flattenPoint + 1 : inlinedSymbols.size();
+    return inlinedSymbols.find_first_in(1, end) == -1;
+  }
+
+  /// Return true if this NLA has a root that originates from a specific module.
+  bool hasRoot(FModuleLike mod) {
+    return (isDead() && nla.root() == mod.moduleNameAttr()) ||
+           rootSet.contains(mod.moduleNameAttr());
+  }
+
+  /// Mark a module as inlined.  This will remove it from the NLA.
+  void inlineModule(FModuleOp module) {
```

Comment:
What if an inner sym needs to be renamed during inlining and is part of the nla?
```
<A::a, B::b>
mod A { inst a of B; inst b of FOO; }
mod B {inst b of BAR;}
```

Reply:
Yeah.... this is currently a problem. The inliner is copying symbols, but not checking for collisions. I was planning to fix this _after_ landing this PR as that bug already existed in the original code.

Reply:
Tracking issue about this: https://github.com/llvm/circt/issues/2451